firstvoices/backend/tests/test_apis/test_immersion_label_api.py

```diff
@@ -0,0 +1,219 @@
+
+    def create_original_instance_for_patch(self, site):
+        return self.create_minimal_instance(site, Visibility.PUBLIC)
+
+    def add_related_objects(self, instance):
+        # no related objects to add
+        pass
+
+    def assert_related_objects_deleted(self, instance):
+        # no related objects to delete
+        pass
+
+    def assert_patch_instance_original_fields(
+        self, original_instance, updated_instance
+    ):
+        assert original_instance.dictionary_entry == updated_instance.dictionary_entry
+
+    def assert_patch_instance_updated_fields(self, data, updated_instance):
+        assert str(updated_instance.key) == data["key"]
+
+    def assert_update_patch_response(self, original_instance, data, actual_response):
+        assert actual_response["key"] == data["key"]
+        assert actual_response["dictionaryEntry"]["id"] == str(
+            original_instance.dictionary_entry.id
+        )
+
+    @pytest.mark.django_db
+    def test_dictionary_entry_same_site_validation(self):
+        """
+        Tests that a validation error is raised if the dictionary entry and immersion label are not in the same site.
+        """
+        user = factories.get_app_admin(AppRole.SUPERADMIN)
+        self.client.force_authenticate(user=user)
+        site = factories.SiteFactory.create()
+        site2 = factories.SiteFactory.create()
+        entry = factories.DictionaryEntryFactory.create(site=site2)
+
+        data = {
+            "dictionary_entry": str(entry.id),
+            "key": self.TEST_KEY,
+        }
+
+        response = self.client.post(
+            self.get_list_endpoint(site_slug=site.slug), format="json", data=data
+        )
+        assert response.status_code == 400
+        assert (
+            response.data["non_field_errors"][0]
+            == "Dictionary entry must belong to the same site as the immersion label."
+        )
+
+    @pytest.mark.django_db
+    def test_immersion_map_permissions(self):
+        """
+        Tests that the immersion map endpoint is only accessible to users with access to the site.
+        """
+        site, user = factories.access.get_site_with_member(
+            Visibility.MEMBERS, Role.MEMBER
+        )
+        user2 = factories.access.get_non_member_user()
+        entry = factories.DictionaryEntryFactory.create(site=site)
+        label = factories.ImmersionLabelFactory.create(
+            site=site, dictionary_entry=entry, key=self.TEST_KEY
+        )
+
+        self.client.force_authenticate(user=user)
+        response = self.client.get(self.get_list_endpoint(site.slug) + "/all")
+
+        assert response.status_code == 200
+        assert response.data[self.TEST_KEY]
+
+        self.client.force_authenticate(user=user2)
+        response = self.client.get(
+            self.get_detail_endpoint(label.key, site.slug) + "all/"
+        )
+        assert response.status_code == 403
+
+    @pytest.mark.django_db
+    def test_immersion_list_empty(self):
```

Comment:
Pretty sure there is a base test case for this already

Reply:
This is testing the list in the `/all` endpoint rather than an empty list on the base endpoint